File: packages/flutterfire_gen_ts/test/helper/chat_message.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    DocumentReference,
    FieldValue,
    QueryDocumentSnapshot,
    QuerySnapshot,
    WriteResult
} from 'firebase-admin/firestore'

export class ReadChatMessage {
  constructor ({
    chatMessageId,
    path,
    senderId,
content,
imageUrls,
isDeleted,
  }: {
    chatMessageId: string
    path: string
    senderId: string
content: string
imageUrls: string[]
isDeleted: boolean
  }) {
    this.chatMessageId = chatMessageId
    this.path = path
    this.senderId = senderId
this.content = content
this.imageUrls = imageUrls
this.isDeleted = isDeleted
  }

  readonly chatMessageId: string

  readonly path: string

  readonly senderId: string

readonly content: string

readonly imageUrls: string[]

readonly isDeleted: boolean

  private static fromJson(json: Record<string, unknown>): ReadChatMessage {
    return new ReadChatMessage({
      chatMessageId: json['chatMessageId'] as string,
      path: json['path'] as string,
      senderId: json['senderId'] as string,
content: (json['content'] as string | undefined) ?? '',
imageUrls: (json['imageUrls'] as (unknown[]  | undefined))?.map((e) => e as string).toList() ?? const <String>[],
isDeleted: (json['isDeleted'] as boolean | undefined) ?? false,
    })
  }


  static fromDocumentSnapshot(
    ds: FirebaseFirestore.DocumentSnapshot
  ): ReadChatMessage {
      const data = ds.data()!
      const cleanedData: Record<string, unknown> = {}
      for (const [key, value] of Object.entries(data)) {
          cleanedData[key] = value === null ? undefined : value
      }
      return ReadChatMessage.fromJson({
          ...cleanedData,
          chatMessageId: ds.id,
          path: ds.ref.path
      })
  }

}

export class CreateChatMessage {
  constructor({
    senderId,
content,
imageUrls,
isDeleted,
  }: {
    senderId: string
content: string
imageUrls: string[]
isDeleted: boolean
  }) {
    this.senderId = senderId
this.content = content
this.imageUrls = imageUrls
this.isDeleted = isDeleted
  }


  readonly senderId: string

readonly content: string

readonly imageUrls: string[]

readonly isDeleted: boolean

  toJson(): Record<string, unknown> {
  return {
    senderId: this.senderId,
content: this.content,
imageUrls: this.imageUrls,
isDeleted: this.isDeleted,
  };
}

}

export class UpdateChatMessage {
  constructor({
    senderId,
content,
imageUrls,
isDeleted,
  }: {
    senderId?: string
content?: string
imageUrls?: string[]
isDeleted?: boolean
  }) {
    this.senderId = senderId
this.content = content
this.imageUrls = imageUrls
this.isDeleted = isDeleted
  }


  readonly senderId?: string

readonly content?: string

readonly imageUrls?: string[]

readonly isDeleted?: boolean

  toJson(): Record<string, unknown> {
  const json: Record<string, unknown> = {}
  if (this.senderId != undefined) {
  json['senderId'] = this.senderId
}
if (this.content != undefined) {
  json['content'] = this.content
}
if (this.imageUrls != undefined) {
  json['imageUrls'] = this.imageUrls
}
if (this.isDeleted != undefined) {
  json['isDeleted'] = this.isDeleted
}

  return json
}

}

export class DeleteChatMessage {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the chatMessages collection for reading.
 */
export const readChatMessageCollectionReference = (
  {
  chatRoomId
}: {
  chatRoomId: string
}
): FirebaseFirestore.CollectionReference<ReadChatMessage>

=> { return db
.collection('chatRooms').doc(chatRoomId).collection('chatMessages').withConverter<ReadChatMessage>({
  fromFirestore: (ds: FirebaseFirestore.DocumentSnapshot): ReadChatMessage => {
    return ReadChatMessage.fromDocumentSnapshot(ds)
  },
  toFirestore: () => {
      throw new Error(`toFirestore is not implemented for ReadChatMessage`)
  }
})
}

/**
 * Provides a reference to a chatMessage document for reading.
 * @param chatMessageId - The ID of the chatMessage document to read.
 */
export const readChatMessageDocumentReference = ({
  chatRoomId,
  chatMessageId,
}: {
  chatRoomId: string,
  chatMessageId: string
}): FirebaseFirestore.DocumentReference<ReadChatMessage> =>
    readChatMessageCollectionReference({
      chatRoomId
    }).doc(chatMessageId);


/**
 * Provides a reference to the chatMessages collection for creating.
 */
export const createChatMessageCollectionReference = (
  {
  chatRoomId
}: {
  chatRoomId: string
}
): FirebaseFirestore.CollectionReference<CreateChatMessage>

=> { return db
.collection('chatRooms').doc(chatRoomId).collection('chatMessages').withConverter<CreateChatMessage>({
  fromFirestore: () => {
    throw new Error(`fromFirestore is not implemented for CreateChatMessage`)
  },
  toFirestore: (obj: CreateChatMessage): FirebaseFirestore.DocumentData => {
      return obj.toJson()
  }
})
}

/**
 * Provides a reference to a chatMessage document for creating.
 * @param chatMessageId - The ID of the chatMessage document to read.
 */
export const createChatMessageDocumentReference = ({
  chatRoomId,
  chatMessageId,
}: {
  chatRoomId: string,
  chatMessageId: string
}): FirebaseFirestore.DocumentReference<CreateChatMessage> =>
    createChatMessageCollectionReference({
      chatRoomId
    }).doc(chatMessageId);


/**
 * Provides a reference to the chatMessages collection for updating.
 */
export const updateChatMessageCollectionReference = (
  {
  chatRoomId
}: {
  chatRoomId: string
}
): FirebaseFirestore.CollectionReference<UpdateChatMessage>

=> { return db
.collection('chatRooms').doc(chatRoomId).collection('chatMessages').withConverter<UpdateChatMessage>({
  fromFirestore: () => {
    throw new Error(`fromFirestore is not implemented for UpdateChatMessage`)
  },
  toFirestore: (obj: UpdateChatMessage): FirebaseFirestore.DocumentData => {
      return obj.toJson()
  }
})
}

/**
 * Provides a reference to a chatMessage document for updating.
 * @param chatMessageId - The ID of the chatMessage document to read.
 */
export const updateChatMessageDocumentReference = ({
  chatRoomId,
  chatMessageId,
}: {
  chatRoomId: string,
  chatMessageId: string
}): FirebaseFirestore.DocumentReference<UpdateChatMessage> =>
    updateChatMessageCollectionReference({
      chatRoomId
    }).doc(chatMessageId);


/**
 * Provides a reference to the chatMessages collection for deleting.
 */
export const deleteChatMessageCollectionReference = (
  {
  chatRoomId
}: {
  chatRoomId: string
}
): FirebaseFirestore.CollectionReference<DeleteChatMessage>

=> { return db
.collection('chatRooms').doc(chatRoomId).collection('chatMessages').withConverter<DeleteChatMessage>({
  fromFirestore: () => {
    throw new Error(`fromFirestore is not implemented for DeleteChatMessage`)
  },
  toFirestore: (): FirebaseFirestore.DocumentData => {
    throw new Error(`toFirestore is not implemented for DeleteChatMessage`)
  }
})
}

/**
 * Provides a reference to a chatMessage document for deleting.
 * @param chatMessageId - The ID of the chatMessage document to read.
 */
export const deleteChatMessageDocumentReference = ({
  chatRoomId,
  chatMessageId,
}: {
  chatRoomId: string,
  chatMessageId: string
}): FirebaseFirestore.DocumentReference<DeleteChatMessage> =>
    deleteChatMessageCollectionReference({
      chatRoomId
    }).doc(chatMessageId);



/**
 * Manages queries against the chatMessages collection.
 */
export class ChatMessageQuery {
  /**
   * Fetches chatMessage documents.
   * @param queryBuilder - Function to modify the query.
   * @param compare - Function to sort the results.
   */
  async fetchDocuments({
      chatRoomId,
      queryBuilder,
      compare
  }: {
      chatRoomId: string,
      queryBuilder?: (
          query: FirebaseFirestore.Query<ReadChatMessage>
      ) => FirebaseFirestore.Query<ReadChatMessage>
      compare?: (lhs: ReadChatMessage, rhs: ReadChatMessage) => number
  }): Promise<ReadChatMessage[]> {
      let query: FirebaseFirestore.Query<ReadChatMessage> =
          readChatMessageCollectionReference({chatRoomId})
      if (queryBuilder != undefined) {
          query = queryBuilder(query)
      }
      const qs: QuerySnapshot<ReadChatMessage> = await query.get()
      let result = qs.docs.map((qds: QueryDocumentSnapshot<ReadChatMessage>) =>
          qds.data()
      )
      if (compare != undefined) {
          result = result.sort(compare)
      }
      return result
  }

  /**
   * Fetches a specific chatMessage document.
   * @param chatMessageId - The ID of the chatMessage document to fetch.
   */
  async fetchDocument({
    chatRoomId,
    chatMessageId
  }: {
    chatRoomId: string,
    chatMessageId: string
  }): Promise<ReadChatMessage | undefined> {
    const ds = await readChatMessageDocumentReference({
      chatRoomId,
      chatMessageId
    }).get()
    return ds.data()
  }

  /**
   * Adds a chatMessage document.
   * @param CreateChatMessage - The chatMessage details to add.
   */
  async add({
    chatRoomId,
    CreateChatMessage
  }: {
    chatRoomId: string,
    CreateChatMessage: CreateChatMessage
  }): Promise<DocumentReference<CreateChatMessage>> {
    return createChatMessageCollectionReference({chatRoomId}).add(CreateChatMessage)
  }

  /**
   * Sets a chatMessage document.
   * @param chatMessageId - The ID of the chatMessage document to set.
   * @param CreateChatMessage - The chatMessage details to set.
   * @param options - Options for the set operation.
   */
  async set({
      chatRoomId,
      chatMessageId,
      CreateChatMessage,
      options
  }: {
      chatRoomId: string,
      chatMessageId: string
      CreateChatMessage: CreateChatMessage
      options?: FirebaseFirestore.SetOptions
  }): Promise<WriteResult> {
      if (options == undefined) {
          return createChatMessageDocumentReference({
            chatRoomId,
            chatMessageId
          }).set(CreateChatMessage)
      } else {
          return createChatMessageDocumentReference({ 
            chatRoomId,
            chatMessageId 
            }).set(CreateChatMessage, options)
      }
  }

  /**
   * Updates a specific chatMessage document.
   * @param chatMessageId - The ID of the chatMessage document to update.
   * @param UpdateChatMessage - The details for updating the chatMessage.
   */
  async update({
    chatRoomId,
    chatMessageId,
    UpdateChatMessage
  }: {
    chatRoomId: string,
    chatMessageId: string
    UpdateChatMessage: UpdateChatMessage
  }): Promise<WriteResult> {
      return updateChatMessageDocumentReference({ 
        chatRoomId,
        chatMessageId 
      }).update(UpdateChatMessage.toJson())
  }

  /**
   * Deletes a specific chatMessage document.
   * @param chatMessageId - The ID of the chatMessage document to delete.
   */
  async delete({
    chatRoomId,
    chatMessageId
  }: {
    chatRoomId: string,
    chatMessageId: string
  }): Promise<WriteResult> {
      return deleteChatMessageDocumentReference({ 
        chatRoomId,
        chatMessageId 
      }).delete()
  }
}

File: packages/flutterfire_gen/test/helper/chat_message.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'chat_message.dart';

class ReadChatMessage {
  const ReadChatMessage({
    required this.senderId,
    required this.map,
    required this.nestedMap,
    required this.deeplyNestedMap,
    required this.chatMessageType,
    required this.content,
    required this.imageUrls,
    required this.isDeleted,
    required this.createdAt,
    required this.updatedAt,
    required this.chatMessageId,
    required this.path,
    required this.chatMessageReference,
  });

  final String senderId;

  final Map<String, dynamic> map;

  final Map<String, Map<String, String>> nestedMap;

  final Map<String, Map<int, List<Map<String, Map<String, dynamic>>>>>
      deeplyNestedMap;

  final ChatMessageType chatMessageType;

  final String content;

  final List<String> imageUrls;

  final bool isDeleted;

  final DateTime? createdAt;

  final DateTime? updatedAt;

  final String chatMessageId;

  final String path;

  final DocumentReference<ReadChatMessage> chatMessageReference;

  factory ReadChatMessage._fromJson(Map<String, dynamic> json) {
    final extendedJson = <String, dynamic>{
      ...json,
    };
    return ReadChatMessage(
      senderId: extendedJson['senderId'] as String,
      map: extendedJson['map'] as Map<String, dynamic>? ??
          const <String, dynamic>{},
      nestedMap: (extendedJson['nestedMap'] as Map<String, dynamic>).map(
          (k, v) => MapEntry(
              k,
              (v as Map<String, dynamic>)
                  .map((k, v) => MapEntry(k, v as String)))),
      deeplyNestedMap: (extendedJson['deeplyNestedMap'] as Map<String, dynamic>)
          .map((k, v) => MapEntry(
              k, v as Map<int, List<Map<String, Map<String, dynamic>>>>)),
      chatMessageType: _chatMessageTypeConverter
          .fromJson(extendedJson['chatMessageType'] as String),
      content: extendedJson['content'] as String? ?? '',
      imageUrls: (extendedJson['imageUrls'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          const <String>[],
      isDeleted: extendedJson['isDeleted'] as bool? ?? false,
      createdAt: (extendedJson['createdAt'] as Timestamp?)?.toDate(),
      updatedAt: (extendedJson['updatedAt'] as Timestamp?)?.toDate(),
      chatMessageId: extendedJson['chatMessageId'] as String,
      path: extendedJson['path'] as String,
      chatMessageReference: extendedJson['chatMessageReference']
          as DocumentReference<ReadChatMessage>,
    );
  }

  factory ReadChatMessage.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadChatMessage._fromJson(<String, dynamic>{
      ...data,
      'chatMessageId': ds.id,
      'path': ds.reference.path,
      'chatMessageReference': ds.reference.parent.doc(ds.id).withConverter(
            fromFirestore: (ds, _) => ReadChatMessage.fromDocumentSnapshot(ds),
            toFirestore: (obj, _) => throw UnimplementedError(),
          ),
    });
  }

  ReadChatMessage copyWith({
    String? senderId,
    Map<String, dynamic>? map,
    Map<String, Map<String, String>>? nestedMap,
    Map<String, Map<int, List<Map<String, Map<String, dynamic>>>>>?
        deeplyNestedMap,
    ChatMessageType? chatMessageType,
    String? content,
    List<String>? imageUrls,
    bool? isDeleted,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? chatMessageId,
    String? path,
    DocumentReference<ReadChatMessage>? chatMessageReference,
  }) {
    return ReadChatMessage(
      senderId: senderId ?? this.senderId,
      map: map ?? this.map,
      nestedMap: nestedMap ?? this.nestedMap,
      deeplyNestedMap: deeplyNestedMap ?? this.deeplyNestedMap,
      chatMessageType: chatMessageType ?? this.chatMessageType,
      content: content ?? this.content,
      imageUrls: imageUrls ?? this.imageUrls,
      isDeleted: isDeleted ?? this.isDeleted,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      chatMessageId: chatMessageId ?? this.chatMessageId,
      path: path ?? this.path,
      chatMessageReference: chatMessageReference ?? this.chatMessageReference,
    );
  }
}

class CreateChatMessage {
  const CreateChatMessage({
    required this.senderId,
    required this.map,
    required this.nestedMap,
    required this.deeplyNestedMap,
    required this.chatMessageType,
    required this.content,
    this.imageUrls = const <String>[],
    this.isDeleted = false,
  });

  final String senderId;
  final Map<String, dynamic> map;
  final Map<String, Map<String, String>> nestedMap;
  final Map<String, Map<int, List<Map<String, Map<String, dynamic>>>>>
      deeplyNestedMap;
  final ChatMessageType chatMessageType;
  final String content;
  final List<String> imageUrls;
  final bool isDeleted;

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'senderId': senderId,
      'map': map,
      'nestedMap': nestedMap,
      'deeplyNestedMap': deeplyNestedMap,
      'chatMessageType': _chatMessageTypeConverter.toJson(chatMessageType),
      'content': content,
      'imageUrls': imageUrls,
      'isDeleted': isDeleted,
      'createdAt': FieldValue.serverTimestamp(),
      'updatedAt': FieldValue.serverTimestamp(),
    };
    final jsonPostProcessors = <({String key, dynamic value})>[];
    return {
      ...json,
      ...Map.fromEntries(jsonPostProcessors
          .map((record) => MapEntry(record.key, record.value))),
    };
  }
}

class UpdateChatMessage {
  const UpdateChatMessage({
    this.senderId,
    this.map,
    this.nestedMap,
    this.deeplyNestedMap,
    this.chatMessageType,
    this.content,
    this.imageUrls,
    this.isDeleted,
    this.createdAt,
  });

  final String? senderId;
  final Map<String, dynamic>? map;
  final Map<String, Map<String, String>>? nestedMap;
  final Map<String, Map<int, List<Map<String, Map<String, dynamic>>>>>?
      deeplyNestedMap;
  final ChatMessageType? chatMessageType;
  final String? content;
  final List<String>? imageUrls;
  final bool? isDeleted;
  final DateTime? createdAt;

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      if (senderId != null) 'senderId': senderId,
      if (map != null) 'map': map,
      if (nestedMap != null) 'nestedMap': nestedMap,
      if (deeplyNestedMap != null) 'deeplyNestedMap': deeplyNestedMap,
      if (chatMessageType != null)
        'chatMessageType': _chatMessageTypeConverter.toJson(chatMessageType!),
      if (content != null) 'content': content,
      if (imageUrls != null) 'imageUrls': imageUrls,
      if (isDeleted != null) 'isDeleted': isDeleted,
      if (createdAt != null) 'createdAt': createdAt,
      'updatedAt': FieldValue.serverTimestamp(),
    };
    final jsonPostProcessors = <({String key, dynamic value})>[];
    return {
      ...json,
      ...Map.fromEntries(jsonPostProcessors
          .map((record) => MapEntry(record.key, record.value))),
    };
  }
}

class DeleteChatMessage {}

/// Provides a reference to the chatMessages collection for reading.
CollectionReference<ReadChatMessage> readChatMessageCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('chatMessages')
        .withConverter<ReadChatMessage>(
          fromFirestore: (ds, _) => ReadChatMessage.fromDocumentSnapshot(ds),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a chatMessage document for reading.
DocumentReference<ReadChatMessage> readChatMessageDocumentReference({
  required String chatRoomId,
  required String chatMessageId,
}) =>
    readChatMessageCollectionReference(chatRoomId: chatRoomId)
        .doc(chatMessageId);

/// Provides a reference to the chatMessages collection for creating.
CollectionReference<CreateChatMessage> createChatMessageCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('chatMessages')
        .withConverter<CreateChatMessage>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a chatMessage document for creating.
DocumentReference<CreateChatMessage> createChatMessageDocumentReference({
  required String chatRoomId,
  required String chatMessageId,
}) =>
    createChatMessageCollectionReference(chatRoomId: chatRoomId)
        .doc(chatMessageId);

/// Provides a reference to the chatMessages collection for updating.
CollectionReference<UpdateChatMessage> updateChatMessageCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('chatMessages')
        .withConverter<UpdateChatMessage>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a chatMessage document for updating.
DocumentReference<UpdateChatMessage> updateChatMessageDocumentReference({
  required String chatRoomId,
  required String chatMessageId,
}) =>
    updateChatMessageCollectionReference(chatRoomId: chatRoomId)
        .doc(chatMessageId);

/// Provides a reference to the chatMessages collection for deleting.
CollectionReference<DeleteChatMessage> deleteChatMessageCollectionReference({
  required String chatRoomId,
}) =>
    FirebaseFirestore.instance
        .collection('chatRooms')
        .doc(chatRoomId)
        .collection('chatMessages')
        .withConverter<DeleteChatMessage>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a chatMessage document for deleting.
DocumentReference<DeleteChatMessage> deleteChatMessageDocumentReference({
  required String chatRoomId,
  required String chatMessageId,
}) =>
    deleteChatMessageCollectionReference(chatRoomId: chatRoomId)
        .doc(chatMessageId);

/// Manages queries against the chatMessages collection.
class ChatMessageQuery {
  /// Fetches [ReadChatMessage] documents.
  Future<List<ReadChatMessage>> fetchDocuments({
    required String chatRoomId,
    GetOptions? options,
    Query<ReadChatMessage>? Function(Query<ReadChatMessage> query)?
        queryBuilder,
    int Function(ReadChatMessage lhs, ReadChatMessage rhs)? compare,
  }) async {
    Query<ReadChatMessage> query =
        readChatMessageCollectionReference(chatRoomId: chatRoomId);
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [ChatMessage] documents.
  Stream<List<ReadChatMessage>> subscribeDocuments({
    required String chatRoomId,
    Query<ReadChatMessage>? Function(Query<ReadChatMessage> query)?
        queryBuilder,
    int Function(ReadChatMessage lhs, ReadChatMessage rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadChatMessage> query =
        readChatMessageCollectionReference(chatRoomId: chatRoomId);
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadChatMessage] document.
  Future<ReadChatMessage?> fetchDocument({
    required String chatRoomId,
    required String chatMessageId,
    GetOptions? options,
  }) async {
    final ds = await readChatMessageDocumentReference(
      chatRoomId: chatRoomId,
      chatMessageId: chatMessageId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [ChatMessage] document.
  Stream<ReadChatMessage?> subscribeDocument({
    required String chatRoomId,
    required String chatMessageId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readChatMessageDocumentReference(
      chatRoomId: chatRoomId,
      chatMessageId: chatMessageId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [ChatMessage] document.
  Future<DocumentReference<CreateChatMessage>> add({
    required String chatRoomId,
    required CreateChatMessage createChatMessage,
  }) =>
      createChatMessageCollectionReference(chatRoomId: chatRoomId)
          .add(createChatMessage);

  /// Sets a [ChatMessage] document.
  Future<void> set({
    required String chatRoomId,
    required String chatMessageId,
    required CreateChatMessage createChatMessage,
    SetOptions? options,
  }) =>
      createChatMessageDocumentReference(
        chatRoomId: chatRoomId,
        chatMessageId: chatMessageId,
      ).set(createChatMessage, options);

  /// Updates a specific [ChatMessage] document.
  Future<void> update({
    required String chatRoomId,
    required String chatMessageId,
    required UpdateChatMessage updateChatMessage,
  }) =>
      updateChatMessageDocumentReference(
        chatRoomId: chatRoomId,
        chatMessageId: chatMessageId,
      ).update(updateChatMessage.toJson());

  /// Deletes a specific [ChatMessage] document.
  Future<void> delete({
    required String chatRoomId,
    required String chatMessageId,
  }) =>
      deleteChatMessageDocumentReference(
        chatRoomId: chatRoomId,
        chatMessageId: chatMessageId,
      ).delete();
}
